examples/seismic/tutorials/06_skimage_tv.ipynb

```diff
@@ -0,0 +1,416 @@
+    "    solver.gradient(rec=residual, u=u0, m=model0.m, grad=grad)\n",
+    "    \n",
+    "    # Taking here to avoid a (probably overzealous) destructor deleting\n",
+    "    # the gradient before Dask has had a chance to communicate it.\n",
+    "    g = np.array(grad.data[:])\n",
+    "    \n",
+    "    # return the objective functional and gradient.\n",
+    "    return fg_pair(f, g)\n",
+    "\n",
+    "import numpy as np\n",
+    "from distributed import Client\n",
+    "\n",
+    "# Dumps the model to disk; workers will pick this up when they need it.\n",
+    "def dump_model(param, model):\n",
+    "    np.save(param['model'], model.astype(np.float32))\n",
+    "\n",
+    "def fwi_gradient(model, param):\n",
+    "    # Dump a copy of the current model for the workers\n",
+    "    # to pick up when they are ready.\n",
+    "    param['model'] = \"model_0.npy\"\n",
+    "    dump_model(param, model)\n",
+    "\n",
+    "    # Define work list\n",
+    "    work = [dict(param) for i in range(param['nshots'])]\n",
+    "    for i in  range(param['nshots']):\n",
+    "        work[i]['shot_id'] = i\n",
+    "        \n",
+    "    # Start dask cluster\n",
+    "    client = Client()\n",
+    "    \n",
+    "    # Map worklist to cluster\n",
+    "    fgi = client.map(fwi_gradient_i, work)\n",
+    "    \n",
+    "    # Perform reduction on cluster\n",
+    "    fg = client.submit(sum, fgi).result()\n",
+    "    \n",
+    "    # Shutdown cluster\n",
+    "    client.close()\n",
+    "    \n",
+    "    # L-BFGS in scipy expects a flat array in 64-bit floats.\n",
+    "    return fg.f, fg.g.flatten().astype(np.float64)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## FWI with L-BFGS-B\n",
+    "Equipped with a function to calculate the functional and gradient, we are finally ready to define the optimization function."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "from scipy import optimize\n",
+    "from skimage.restoration import denoise_tv_chambolle as denoise\n",
+    "\n",
+    "# Define bounding box constraints on the solution.\n",
+    "def apply_box_constraint(m):\n",
+    "    # Maximum accepted velocity is 3.5 km/sec (true is 3 km/sec)\n",
+    "    # Minimum accepted velocity is 2 km/sec (true is 2.5 km/sec)\n",
+    "    return np.clip(m, 0.08, 0.25)\n",
+    "\n",
+    "# Many methods in scipy.optimize.minimize allow us to define a callback\n",
+    "# function that can operate on the solution after every iteration. Here\n",
+    "# we use this to apply box constraints and to monitor the true relative\n",
+    "# solution error.\n",
+    "relative_error = []\n",
+    "\n",
+    "def fwi_tv_callbacks(x):\n",
+    "    # Apply boundary constraint\n",
+    "    x.data[:] = denoise(x.reshape(181, 181), weight=5.0e-3).flatten().data\n",
```

Comment:
Why do I need a `.data` at the end of this line?

Reply:
fixed